pyresample/test/test_spherical_utils.py

```diff
@@ -0,0 +1,361 @@
+                         [25, 40, 40, 25]]).T
+    poly2 = SphPolygon(np.deg2rad(vertices))
+
+    res = check_if_two_polygons_overlap(poly1, poly2)
+
+    assert res is False
+
+
+@patch.object(GetNonOverlapUnionsBaseClass, '_merge_unions', fake_merge_unions)
+@patch('pyresample.spherical_utils.check_keys_int_or_tuple')
+def test_merge_when_input_objects_do_not_overlap(check_keys_int_or_tuple):
+    """Test main method (merge) of the GetNonOverlapUnionsBaseClass class."""
+
+    mysets = [{1, 3, 5, 7, 9}, {2, 4, 6, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19}]
+    myobjects = GetNonOverlapUnionsBaseClass(mysets)
+
+    check_keys_int_or_tuple.return_code = None
+
+    with patch('pyresample.spherical_utils.merge_tuples', return_value=0):
+        myobjects.merge()
+
+    polygons = myobjects.get_polygons()
+
+    assert polygons == mysets
+
+
+@patch.object(GetNonOverlapUnionsBaseClass, '_merge_unions', fake_merge_unions)
+@patch('pyresample.spherical_utils.check_keys_int_or_tuple')
+def test_merge_overlapping_and_nonoverlapping_objects(check_keys_int_or_tuple):
+    """Test main method (merge) of the GetNonOverlapUnionsBaseClass class."""
+    mysets = [SET_A, SET_B, SET_C, SET_D, SET_E, SET_F, SET_G]
+    myobjects = GetNonOverlapUnionsBaseClass(mysets)
+
+    check_keys_int_or_tuple.return_code = None
+
+    with patch('pyresample.spherical_utils.merge_tuples') as mypatch:
+        mypatch.side_effect = fake_merge_tuples
+        myobjects.merge()
+
+    polygons = myobjects.get_polygons()
+    ids = myobjects.get_ids()
+
+    expected = [{1, 3, 5, 7, 9},
+                {2, 4, 6, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19},
+                {20, 21, 22, 23, 24, 26, 27, 28, 29}]
+
+    assert polygons == expected
+
+    expected = [0, (2, 1, 3), (5, 4, 6)]
+
+    assert ids == expected
+
+
+def test_flatten_tuple():
+    """Test flatten a nested tuple of integers"""
+
+    with pytest.raises(TypeError) as exec_info:
+        _ = merge_tuples(7)
+
+    exception_raised = exec_info.value
+    assert str(exception_raised) == "Function argument must be a tuple!"
+
+    intuple = (((0, 1), (2, 3)), ((4, 5), (6, 7)))
+
+    with patch('pyresample.spherical_utils.int_items_to_tuples') as mypatch:
+        mypatch.side_effect = fake_int_items_to_tuples
+        res = merge_tuples(intuple)
+
+    assert res == (0, 1, 2, 3, 4, 5, 6, 7)
+
+    intuple = (3, (1, 2))
+    with patch('pyresample.spherical_utils.int_items_to_tuples') as mypatch:
+        mypatch.side_effect = fake_int_items_to_tuples
+        res = merge_tuples(intuple)
+
+    assert res == (3, 1, 2)
+
+    intuple = (0, (1, 2), (3, (4, 5)))
+    with patch('pyresample.spherical_utils.int_items_to_tuples') as mypatch:
+        mypatch.side_effect = fake_int_items_to_tuples
+        res = merge_tuples(intuple)
+
+    assert res == (0, 1, 2, 3, 4, 5)
+
+    intuple = (0, (1, 2), (3, (4, 5)), (6,))
+    with patch('pyresample.spherical_utils.int_items_to_tuples') as mypatch:
```

Comment:
This is a lot of mocking. Is it possible to write real polygon objects that test these same cases without needing to mock?

Reply:
Here I just delegated some very small task to the function `int_items_to_tuples` and avoiding testing that too when testing the behaviour of `merge_tuples`. There are no polygons involved. Most of the code in `spherical_utils` is actually independent from geometric polygons.

Reply:
How hard would it be to do these tests with no mocking?

Reply:
@djhoese I will reduce the mocking. Talked to @mraspaud about testing and I realize I have been obsessed lately with reducing my tests to not test functions inside functions, etc. I realize now that actually makes my tests more dependent on the implementation than the opposite!

Reply:
Exactly my concern!

Reply:
I skipped some of the mocking at least. Better now?

Reply:
And sorry for the long delay @djhoese